vortex-array/src/builders/primitive.rs

```diff
@@ -48,23 +49,50 @@ impl<T: NativePType> PrimitiveBuilder<T> {
     }
 
     pub fn patch(&mut self, patches: Patches, starting_at: usize) -> VortexResult<()> {
-        let (array_len, offset, indices, values) = patches.into_parts();
+        let (array_len, indices_offset, indices, values) = patches.into_parts();
         assert!(starting_at + array_len == self.len());
         let indices = indices.into_primitive()?;
         let values = values.into_primitive()?;
         let validity = values.validity_mask()?;
         let values = values.as_slice::<T>();
         match_each_unsigned_integer_ptype!(indices.ptype(), |$P| {
-            for (compressed_index, decompressed_index) in indices.into_primitive()?.as_slice::<$P>().into_iter().enumerate() {
-                let decompressed_index = *decompressed_index as usize;
-                let out_index = starting_at + decompressed_index - offset;
-                if validity.value(compressed_index) {
-                    self.values[out_index] = values[compressed_index]
-                } else {
-                    self.nulls.set_bit(out_index, false)
-                }
+            if !matches!(validity, Mask::AllTrue(_)) {
```

Comment:
Sorry to be annoying but can we move this into either two matches or an inner fn? @0ax1 noticed that this match + more inner code can affect perf opts from the compiler

Reply:
Also do you think its a good idea to implement this with two loops in the nullable case?

Reply:
If code size is no concern, I think the best code is: switch on nullability, in the nullable case, have one loop over cache-sized chunks of indices with two inner loops that update the nulls and the values.

I could benchmark these approaches. I mostly felt the effect wouldn't be that large and I'm not sure of a way to pick the right size of the chunks. Patches should usually be a pretty small fraction of the array.

Reply:
I benchmarked two loops vs one loop for nullable, and your hunch was right that one loop is best.